src/functional/fencil_processors/processor_interface.py

```diff
@@ -0,0 +1,139 @@
+    @property
+    def kind(self) -> type[FencilFormatter]:
+        return FencilFormatter
+
+
+def fencil_formatter(
+    func: FencilProcessorFunction[str],
+) -> FencilProcessorProtocol[str, FencilFormatter]:
+    """
+    Turn a formatter function into a FencilFormatter.
+
+    Examples:
+    ---------
+    >>> @fencil_formatter
+    ... def format_foo(fencil: FencilDefinition, *args, **kwargs) -> str:
+    ...     '''A very useless fencil formatter.'''
+    ...     return "foo"
+
+    >>> ensure_processor_kind(format_foo, FencilFormatter)
+    """
+    # this operation effectively changes the type of func and that is the intention here
+    func.kind = FencilFormatter  # type: ignore[attr-defined]
+    return cast(FencilProcessorProtocol[str, FencilFormatter], func)
+
+
+class FencilSourceModuleGenerator(
+    FencilProcessorProtocol[SourceModule, "FencilSourceModuleGenerator"], Protocol
+):
+    @property
+    def kind(self) -> type[FencilSourceModuleGenerator]:
+        return FencilSourceModuleGenerator
+
+
+def fencil_source_module_generator(
+    func: FencilProcessorFunction[SourceModule],
+) -> FencilProcessorProtocol[SourceModule, FencilSourceModuleGenerator]:
+    """
+    Wrap a source module generator function in a ``FencilSourceModuleGenerator`` instance.
+
+    Examples:
+    ---------
+    >>> from .source_modules.source_modules import Function
+    >>> @fencil_source_module_generator
+    ... def generate_foo(fencil: FencilDefinition, *args, **kwargs) -> SourceModule:
+    ...     '''A very useless fencil formatter.'''
+    ...     return SourceModule(entry_point=Function(fencil.id, []), library_deps=[], source_code="foo", language="foo")
+
+    >>> ensure_processor_kind(generate_foo, FencilSourceModuleGenerator)
+    """
+    # this operation effectively changes the type of func and that is the intention here
+    func.kind = FencilSourceModuleGenerator  # type: ignore[attr-defined]
+    return cast(FencilProcessorProtocol[SourceModule, FencilSourceModuleGenerator], func)
+
+
+class FencilExecutor(FencilProcessorProtocol[None, "FencilExecutor"], Protocol):
+    @property
+    def kind(self) -> type[FencilExecutor]:
+        return FencilExecutor
+
+
+def fencil_executor(
+    func: FencilProcessorFunction[None],
+) -> FencilProcessorProtocol[None, FencilExecutor]:
+    """
+    Wrap an executor function in a ``FencilFormatter`` instance.
+
+    Examples:
+    ---------
+    >>> @fencil_executor
+    ... def badly_execute(fencil: FencilDefinition, *args, **kwargs) -> None:
+    ...     '''A useless and incorrect fencil executor.'''
+    ...     pass
+
+    >>> ensure_processor_kind(badly_execute, FencilExecutor)
+    """
+    # this operation effectively changes the type of func and that is the intention here
+    func.kind = FencilExecutor  # type: ignore[attr-defined]
+    return cast(FencilProcessorProtocol[None, FencilExecutor], func)
+
+
+def is_processor_kind(
+    obj: object, kind: type[ProcessorKindT]
```

Comment:
Nice trick to define a parametric typeguard! I'm wondering though, if it shouldn't contain also the `OutputT` also as type var of `FencilProcessorProtocol` to be able to narrow it down properly

```suggestion
    obj: FencilProcessorProtocol[OutputT, Any?? ProcessorKindT?? ...??], kind: type[ProcessorKindT]
```

Reply:
If you want to do that, then we have the `__name__` problem again. Because (quote mypy):
>FencilProcessorProtocol[Any, Any] does not have a __name__ attribute

Reply:
Also, this is basically only useful at runtime, where the output type would be completely ignored.

Reply:
Ah, wait, we don't need the `__name__` anymore.

Reply:
I added it but mind you, it is purely cosmetic since the `OutputT` only occurs that one time and doesn't have to match up with anything. Basically it is completely generic on `OutputT` with no constraints whatsoever.

Reply:
But I think it appears twice  in the `is_processor_kind` typeguard: once in the typing of the first argument and the other time in the `TypeGuard[]` result type, right?